Add Flax FAQ - how to search, how to take the derivative w.r.t. a hidden layer, remat_scan vs scan(remat), recommended training libraries/metrics (#3267)
google/flax: docs/faq.rst

Frequently Asked Questions (FAQ)
================================

This is a collection of answers to frequently asked questions (FAQ). You can contribute to the Flax FAQ by starting a new topic in `GitHub Discussions <https://github.com/google/flax/discussions>`__.

Where to search for an answer to a Flax-related question
********************************************************

There are a number of official Flax resources to search for information:

> - `Flax Documentation on ReadTheDocs <https://github.com/google/flax/discussions>`__ (this site): Use the search bar or the table of contents on the left-hand side.

Reviewer: ```suggestion
- `Flax Documentation on ReadTheDocs <https://flax.readthedocs.io/en/latest/>`__ (this site): Use the `search bar <https://flax--3267.org.readthedocs.build/en/3267/search.html>` or the table of contents on the left-hand side.
```

Author: :+1: 
Nit: https://flax.readthedocs.io/en/latest/search.html

- `google/flax GitHub Discussions <https://github.com/google/flax/discussions>`__: Search for an existing topic or start a new one. If you can't find what you're looking for, feel free to ask the Flax team or community a question.

> - `google/flax GitHub Issues <https://github.com/google/flax/discussions>`__.

Reviewer: Also maybe add a brief description for searching in Github Issues

> How to take the derivative with respect to a hidden layer values (using :code:`Module.perturb`)

Reviewer 2: ```suggestion
How to take the derivative with respect to an intermediate value (using :code:`Module.perturb`)
```

***********************************************************************************************

To take the derivative(s) or gradient(s) of the output with respect to a hidden/intermediate activation inside a model layer, you can use :meth:`flax.linen.Module.perturb`. You define a zero-value :class:`flax.linen.Module` "perturbation" parameter – :code:`perturb(...)` – in the forward pass with the same shape as the intermediate activation, define the loss function with :code:`'perturbations'` as an added standalone argument, perform a JAX derivative operation with :code:`jax.grad` on the perturbation argument.

For full examples and detailed documentation, go to:

- The :meth:`flax.linen.Module.perturb` API docs
- The `Extracting gradients of intermediate values <https://flax.readthedocs.io/en/latest/guides/extracting_intermediates.html#extracting-gradients-of-intermediate-values>`_ guide
- `Flax GitHub Discussions #1152 <https://github.com/google/flax/discussions/1152>`_

Is Flax Linen :code:`remat_scan()` the same as :code:`scan(remat(...))`?
************************************************************************

> Flax :code:`remat_scan()` (:meth:`flax.linen.remat_scan()`) and :code:`scan(remat(...))` (:meth:`flax.linen.scan` over :meth:`flax.linen.remat`) are not the same, and :code:`remat_scan()` is limited in cases it supports.

Reviewer 2: Here we should clarify what "is limited in cases it supports" means, specifically that `remat_scan` treats the inputs and outputs as carries (there are no scan dimension for the data). This makes it easy to implement the "scan over layers" pattern for an existing Module as shown in the docs for `remat_scan`, with `scan(remat(...))` you have create a wrapper Module to handle the scan inputs/outputs.

You are recommended to use :code:`scan(remat(...))`, as typically you would need the extra parameters, such as ``in_axes`` or ``out_axes``, which :meth:`flax.linen.remat_scan` does not expose.

> References:

Reviewer: Might not need these links below since they were already mentioned previously

- The :meth:`flax.linen.Module.perturb` API docs
- The `Extracting gradients of intermediate values <https://flax.readthedocs.io/en/latest/guides/extracting_intermediates.html#extracting-gradients-of-intermediate-values>`_ guide
- `Flax GitHub Discussions #1152 <https://github.com/google/flax/discussions/1152>`__

Is Flax Linen :code:`remat_scan()` the same as :code:`scan(remat(...))`?
************************************************************************

Flax :code:`remat_scan()` (:meth:`flax.linen.remat_scan()`) and :code:`scan(remat(...))` (:meth:`flax.linen.scan` over :meth:`flax.linen.remat`) are not the same, and :code:`remat_scan()` is limited in cases it supports. Namely, :code:`remat_scan()` treats the inputs and outputs as carries (hidden states that are carried through the training loop). You are recommended to use :code:`scan(remat(...))`, as typically you would need the extra parameters, such as ``in_axes`` (for input array axes) or ``out_axes`` (output array axes), which :meth:`flax.linen.remat_scan` does not expose.

What are the recommended training loop libraries?
*************************************************

> Consider using CLU (Common Loop Utils) `google/CommonLoopUtils <https://github.com/google/CommonLoopUtils>`__. To get started, go to this `CLU Synopsis Colab <https://colab.research.google.com/github/google/CommonLoopUtils/blob/main/clu_synopsis.ipynb>`__. You can find answers to common questions about CLU with Flax on `google/flax GitHub Discussions <https://github.com/google/flax/discussions>`__.

Reviewer: ```suggestion
Consider using CLU (Common Loop Utils) `google/CommonLoopUtils <https://github.com/google/CommonLoopUtils>`__. To get started, go to this `CLU Synopsis Colab <https://colab.research.google.com/github/google/CommonLoopUtils/blob/main/clu_synopsis.ipynb>`__. You can find answers to common questions about CLU with Flax on `google/flax GitHub Discussions <https://github.com/google/flax/discussions?discussions_q=clu>`__.
```
Maybe link it to discussions filtered for CLU?
